### PR TITLE
check the trainable type for Layer init

### PR DESCRIPTION
**Pull request by haifeng-jin**
Enforce the `trainable` arg in `Layer.__init__` to be boolean.
It seems in the tests there are some use cases setting `trainable=None`.

### CODE REVIEW

File: keras/engine/base_layer.py [hunk 1]
```diff
@@ -325,6 +325,8 @@ def __init__(self,
     # Mutable properties
     # Indicates whether the layer's weights are updated during training
     # and whether the layer's updates are run during training.
+    if not isinstance(trainable, bool):
+      raise TypeError('Expect trainable to be a boolean, got ' + str(trainable))
```
Comment (fchollet) [inline keras/engine/base_layer.py:329]: Use a f-string for string formatting

File: keras/engine/base_layer.py [hunk 2]
```diff
@@ -325,6 +325,8 @@ def __init__(self,
     # Mutable properties
     # Indicates whether the layer's weights are updated during training
     # and whether the layer's updates are run during training.
+    if not isinstance(trainable, bool):
```
Comment (fchollet) [inline keras/engine/base_layer.py:328]: The `trainable` argument should also support a scalar boolean `tf.Variable`.

In addition, there is a risk here that some people might be passing `0` or `1` and would then see their code break. We would need to run a global TAP to check for such cases.

File: keras/engine/base_layer_test.py [hunk 3]
```diff
@@ -462,6 +462,11 @@ def __init__(self):
     self.assertEqual(self.evaluate(layer.weights[0]), 1.)
     self.assertEqual(self.evaluate(layer.weights[1]), 2.)
 
+  def test_exception_if_trainable_not_boolean(self):
+    with self.assertRaises(TypeError) as e:
```
Comment (fchollet) [inline keras/engine/base_layer_test.py:466]: Any `assertRaises` should use a regex to check that the error being raised is the one you expect (and not another error upstream)

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Review by fchollet — APPROVED** [review]
LGTM, thanks. When merging we'll need to patch the CL and run a global TAP.

**Comment by haifeng-jin** [comment]
@fchollet One more commit for using the regex to match the error message is in.

**Comment by haifeng-jin** [comment]
Will submit internally.

**Pull request closed without merging**